### ISSUE TITLE
Pass arguments from the command line into a single configurable recipe

### ISSUE THREAD

**Issue by commonquail** [openrewrite/rewrite-maven-plugin#799]
We can already run OpenRewrite via the Maven plugin without modifying files up front:

```sh
mvn -U org.openrewrite.maven:rewrite-maven-plugin:run \
  -Drewrite.activeRecipes=org.openrewrite.java.RemoveUnusedImports
```

but sadly [only for recipes that require no configuration][:(]. Recipes that do require configuration have no (canonical) inline invocation form, instead necessitating the creation of a verbose and complex file merely to communicate what is often trivial information to the recipe, and then you even _still_ need the `rewrite.activeRecipes` user property from the example. This severely limits the practical utility of OpenRewrite. Instead of being a hypothetical arbitrary rename-refactor the tool is only really interesting for completely static transformations, or for dynamic transformations that are particularly complex or that have to run in batch mode.

[:(]: https://docs.openrewrite.org/running-recipes/running-rewrite-on-a-maven-project-without-modifying-the-build#running-a-recipe-with-configuration-parameters

## What problem are you trying to solve?

Consider a trivial recipe with a structure comparable to

```yaml
---
type: specs.openrewrite.org/v1beta/recipe
name: com.yourorg.ChangePackageExample
displayName: Rename package name example
recipeList:
  - org.openrewrite.java.ChangePackage:
      oldPackageName: com.yourorg.foo
      newPackageName: com.yourorg.bar
```

Suppose this needs to be run by _many different individuals_ for different values of `oldPackageName` or `newPackageName`. For \<reasons> we can distribute instructions for defining and executing a recipe but distributing a recipe itself is impractical; say, for example, that we cannot know literally every variation of package names, or that the old package name can exist in multiple artifacts and should be turned into numerous divergent package names.

The YAML configuration file works well with distribution when distribution is already a solved problem, but distribution is a difficult problem to solve and cumbersome even when already solved. It requires a distribution channel, which you may not already have, and if you do have one, using it may necessitate a disproportionate amount of bureaucracy namely if you're subject to auditing regulations. Even if distribution is feasible, some recipes are impossible to distribute in practice, like the examples from the previous paragraph.

But I don't see why we couldn't "just" pass the package names as user properties like we can with `rewrite.activeRecipes`.

## Describe the solution you'd like

Imagine I could run something like

```sh
mvn -U org.openrewrite.maven:rewrite-maven-plugin:run \
  -Drewrite.activeRecipes=org.openrewrite.java.ChangePackage \
  -Drewrite.ChangePackage.oldPackageName=com.yourorg.foo \
  -Drewrite.ChangePackage.newPackageName=com.yourorg.bar
```

Generalizing, `-Drewrite.RecipeSimpleName.optionName=value`.

This is a command I can email blast to lots of individuals, each of whom can easily adjust the invocation as necessary with no further effort.

The pattern I propose means you can run each distinct recipe only once per invocation but is acceptable to me. Repeating the command for variations in configuration option values is trivial, and if you need to do that a lot you're probably in a situation where the overhead of the configuration file approach is amortized anyway. An _n_-ary invocation form is theoretically possible, it just doesn't seem worth the implementation complexity or the impact to invocation complexity.

I don't care about the exact format of the user property. It only needs to be reasonably resistant to collision and should feature the option name verbatim for clarity.

## Have you considered any alternatives or workarounds?

On multiple occasions I've discounted OpenRewrite in favour of low level tools like `sed` and `mv` and Python on account of this limitation. I've thought about making tools to generate specialized OpenRewrite configuration files but that has not been worth the effort so far.

**Comment by timtebeek** [comment]
We've discussed this some more internally @commonquail ; if you're up for it we'd welcome a contribution to take in command line arguments and pass them into a single configurable recipe. We could then incubate this for a while and see how that works out in practice.

**Comment by commonquail** [comment]
> Have you thought of going the opposite direction? Not send out the commands for them to run, but the results of those commands in a PR?

Indeed. Working on behalf of people is not an option either, for bureaucratic or practical reasons or both. I mean, it's all a bit inefficient and error prone, to be sure, but these are the constraints.

> That might indeed be the case; it's always difficult to visualize a simplified and redacted use case, and I can understand that you can't share everything you're hoping to use these tools for.

I didn't mean to sound secretive there, it's just that if the recipe's behaviour is necessarily contextual then it's just not possible to centralize the recipe (short of creating _n_ concretizations of the same template recipe but in practice that's less feasible than written instructions).

> we'd welcome a contribution to take in command line arguments and pass them into a single (non-composite) configurable recipe.

Cool! I can't make any commitments but it's nice to know you'd be willing to evaluate the functionality.

> I've moved the issue as this would be unique to the Maven plugin

Thanks. I personally wouldn't have been able to help with Gradle in any case.

**Comment by marvin-froeder** [comment]
Any updates on this?

I'm on a somewhat airgap environment, where I can reach maven central (via proxy) but no way I can use `moderne-cli` as it is too reliant on external app

**Comment by timtebeek** [comment]
Just to be clear: the Moderne CLI has no reliance on any external app; it can work fully locally if you want it to, without any connection to the Moderne platform or DX. We'll use the exact same configuration as Maven does when connecting through a proxy to Maven Central. Welcome to [give the CLI a try](https://www.moderne.io/moderne-ide-plugin-signup).

Other than that the offer from above still stands: if anyone wants to contribute similar argument handling for standalone recipes to the rewrite-maven-plugin, then we'll review and maintain that going forward.

**Comment by marvin-froeder** [comment]
Interesting, cause I couldn't move forward with the client without running

`mod config moderne edit https://app.moderne.io --token mat-YOUR_TOKEN_HERE` 

but that gave me an error due to airgap env

**Comment by timtebeek** [comment]
You'll likely want to use `mod config license edit YOUR_TOKEN` instead, as described here:
https://docs.moderne.io/user-documentation/moderne-cli/getting-started/moderne-cli-license#everyone-else-including-multi-tenant-moderne-customers

In short we have various ways for folks to run the Moderne CLI, and _optionally_ hook that up to a in-house or hosted Moderne instance. You've followed the instructions for folks that have their private instance of Moderne, whereas you'll want to follow the flow "for everyone else" as outlined in the docs.

Hope that helps you get started for now! For other folks: you can [get a license through this form](https://www.moderne.io/moderne-ide-plugin-signup).

**Comment by thombergs** [comment]
I'm very much interested in this as well, as we're currently trying to leverage the Maven plugin to (automatically) execute arbitrary recipes from the command line without having to create `rewrite.yml` files on the fly.

I'll take a look at the code next week to see if I can contribute this and may come back with questions here :)

**Comment by timtebeek** [comment]
Thanks for reaching out @thombergs ! Note that there's an ongoing effort in this PR, with some small steps left to do:
- https://github.com/openrewrite/rewrite-maven-plugin/pull/816

You'd already be able to use the same in the Moderne CLI, for folks who have access to that. Let me know if you'd like help to set that up.

**Comment by timtebeek** [comment]
- https://github.com/openrewrite/rewrite-maven-plugin/pull/816 is now ready for review; I welcome your feedback before a merge.
